Cortex Generator Support

SchrodingersGat opened the pull request:
This PR adds support for the [Cortex Hybrid Power System](https://www.currawongeng.com/chps/) (CHPS) generator from Currawong Engineering.

The generator supported via PiccoloCAN protocol, and this PR supports:

- Telemetry data
- Mavlink integration
- Remote engine starting

SchrodingersGat commented:
@tridge who best to review this?

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.cpp:
```diff
@@ -0,0 +1,299 @@
+{
+    const uint32_t now = AP_HAL::millis();
+
+    return (now - last_reading_ms) < 2000;
```
2 seconds seems *way* too generous!

SchrodingersGat replied:
Is there a standard or default timeout value to be used here? Otherwise, what would you suggest I reduce it to?

peterbarker replied:
No - how long do you think it reasonable that the generator just simply disappear for? :-)

I'm responsible for the existing timeouts being as long as they are - and for the life of me I don't understand how 2 seconds is reasonable on the RichenPower, for example.

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.cpp:
```diff
@@ -0,0 +1,299 @@
+    // Send the "standby" command to the Cortex generator
+    // Note: If the engine is not also disabled, the generator may restart automatically
+
+    AP_HAL::CANFrame txFrame {};
```
Move this to be just above where its content is filled in.

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.cpp:
```diff
@@ -0,0 +1,299 @@
+
+bool AP_Generator_Cortex::run(void)
+{
+    AP_HAL::CANFrame txFrame {};
```
Move this down to where it's initialised

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.h:
```diff
@@ -0,0 +1,109 @@
+    }
+
+private:
+    bool handle_message(AP_HAL::CANFrame &frame, AP_PiccoloCAN* can_iface);
```
If there's never a case that `can_iface` can come in as nullptr then pass it as a reference (and take its address as you need to)

peterbarker commented on libraries/AP_Generator/AP_Generator_config.h:
```diff
@@ -28,4 +28,8 @@
+#endif
+
+#ifndef AP_GENERATOR_CORTEX_ENABLED
+#define AP_GENERATOR_CORTEX_ENABLED AP_GENERATOR_BACKEND_DEFAULT_ENABLED && HAL_PROGRAM_SIZE_LIMIT_KB > 1024
```
No generator is included by default, on the basis that about 0.001% of our users will actually be flying generators, and since we're tight on space we have to leave some stuff out by default.

Our [custom build server](https://custom.ardupilot.org/) can be used to add the feature to a build for those that need something not compiled in by default.

OTOH, someone who is flying a vehicle serious enough to swing a generator around should probably be running an [OEM setup](https://ardupilot.org/copter/docs/common-oem-customizations.html) where they're stashing all the customisation typically required for a larger workhorse aircraft.

You can also simply enable it on the `waf` command-line (e.g.):
```
pbarker@crun:~/rc/ardupilot(pr/remove-thread-directives)$ ./waf --help | grep -i loweheiser
    --enable-GENERATOR_LOWEHEISER
                        Enable Loweheiser generator
    --disable-GENERATOR_LOWEHEISER
                        Disable Loweheiser generator
pbarker@crun:~/rc/ardupilot(pr/remove-thread-directives)$ 
```

SchrodingersGat commented:
@peterbarker I have rebased the PR, squashed commits and updated the commit messages. Please let me know if there are any other changes you would like to see here

peterbarker commented:
> From one of the failing builds:
> 
> ```
> ValueError: feature (AP_GENERATOR_CORTEX) is not matched in extract_features
> ```
> 
> Any suggestions how to address this?

You need to add a line to `extract_features.py` so that we can tell when this feature is built into ArduPilot.  We use this to pre-populate fields on the custom-build-server, for example.

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.cpp:
```diff
@@ -0,0 +1,352 @@
+                }
+            break;
+        default:
+            break;
```
```suggestion
```

if you sprout more modes you should probably be forced to handle them here :-)

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.cpp:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    return true;
+}
```
Maybe just put it in the header and `return connected`?

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.cpp:
```diff
@@ -0,0 +1,352 @@
+#else
+    // PiccoloCAN interface not supported
+    return false;
+#endif // HAL_PICCOLO_CAN_ENABLE
```
```suggestion
```

... and everywhere else.  Just assume in this file that `HAL_PICCOLO_CAN_ENABLE` is set

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.h:
```diff
@@ -0,0 +1,127 @@
+
+#include "AP_Generator_Backend.h"
+#include <AP_Common/AP_Common.h>
+#include <AP_Logger/AP_Logger_config.h>
```
```suggestion
#include <AP_Common/AP_Common.h>
```

SchrodingersGat replied:
Cleaned

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.h:
```diff
@@ -0,0 +1,127 @@
+#include <AP_PiccoloCAN/AP_PiccoloCAN_Device.h>
+#include <AP_PiccoloCAN/piccolo_protocol/CortexPackets.h>
+#include <stdint.h>
+#include <stdio.h>
```
```suggestion
```

shouldn't need stdio.

We omit stdint even 'though it's probably strictly required.

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.h:
```diff
@@ -0,0 +1,127 @@
+private:
+    static AP_Generator_Cortex* _singleton;
+
+#if HAL_PICCOLO_CAN_ENABLE
```
```suggestion
```

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.h:
```diff
@@ -0,0 +1,127 @@
+    uint16_t _can_device_type = 0;
+    uint8_t _can_device_id = 0xFF;
+
+    friend class AP_PiccoloCAN;
```
Do you really need to friend that?!

peterbarker reviewed (changes requested):
@SchrodingersGat there's still some comments from my last PR left un-addressed.

Please also fix the commit list - we should have just three - one for `Tools:`, one for `AP_PiccoloCAN`, one for `AP_Generator`.

I'd have fixed the commits and merged if my nit-picky review comments had been address :-P

SchrodingersGat commented:
@peterbarker I believe that this now has addressed all the issues, and I have compressed into three separate commits as requested. Fingers crossed this is the one!

peterbarker commented on libraries/AP_Generator/AP_Generator_Cortex.cpp:
```diff
@@ -0,0 +1,294 @@
+
+    // Return the maximum of the two internal temperature values
+    return controller.rectifierTemperature > controller.regulatorTemperature ?
+        controller.rectifierTemperature : controller.regulatorTemperature;
```
```suggestion
    return MAX(controller.rectifierTemperature, controller.regulatorTemperature);
```
